### PR TITLE
Add overflow handling to `ButtonGroup`

### LINKED ISSUE #196: Overflow button group
#### Feature
If the toolbar grows beyond it’s available horizontal or vertical space, we place the tools within an overflow button. When the overflow button is selected, an additional toolbar slides perpendicular away from the overflow button.

#### Examples
image

### PR DESCRIPTION
**Pull request by mayank99**
Extracted out overflow logic from `Breadcrumbs` into `useOverflow` hook.

Added `overflowButton` prop to `ButtonGroup` to allow rendering a custom dropdown as last button.

Closes #196

image

## Checklist

- [x] Add meaningful unit tests for your component (verify that all lines are covered)
- [x] Verify that all existing tests pass
- [x] Add component features demo in Storybook (different stories)
- [x] Approve test images for new stories
- [ ] Add an entry to the changelog for any new components and changes
- [x] Add screenshots of the key elements of the component
- [x] Add any additional comments describing the features you've implemented

### CODE REVIEW

File: src/core/ButtonGroup/ButtonGroup.tsx [hunk 1]
```diff
@@ -4,16 +4,21 @@
...
 export type ButtonGroupProps = {
   /**
    * Buttons in the ButtonGroup.
    */
   children: React.ReactNode;
-} & StylingProps;
+  /**
+   * If true, the ButtonGroup will collapse the overflowing Buttons into a menu.
+   */
+  responsive?: boolean;
```
Comment (bentleyvk) [inline src/core/ButtonGroup/ButtonGroup.tsx:20]: We might also need some callback in order for users to be able to show hidden items in a dropdown menu. Or should we handle that?

Comment (mayank99) [reply src/core/ButtonGroup/ButtonGroup.tsx:20]: Oh this is not complete. I wanted to fix the hook/test first.

I plan on adding:
- "..." as the last icon
- excessIconProps
- dropdownItems

Comment (mayank99) [reply src/core/ButtonGroup/ButtonGroup.tsx]: ok, all of the above are now supported using the `overflowButton` prop.

File: src/core/ButtonGroup/ButtonGroup.tsx [hunk 2]
```diff
@@ -4,19 +4,30 @@
...
+   * Expects a function that takes the index of the first button that is overflowing (i.e. hidden)
+   * and returns the `ReactNode` to render.
+   */
+  overflowButton?: (firstOverflowingIndex: number) => React.ReactNode;
+} & Omit<CommonProps, 'title'>;
 
 /**
- * Group buttons together for common actions
+ * Group buttons together for common actions.
+ * Handles reponsive overflow when the `overflowButton` prop is specified.
```
Comment (bentleyvk) [inline src/core/ButtonGroup/ButtonGroup.tsx:29]: reSponsive

File: stories/core/Buttons/ButtonGroup.stories.tsx [hunk 3]
```diff
@@ -37,3 +54,61 @@ export const WithIcons: Story<ButtonGroupProps> = () => {
...
+  return (
+    <div
+      style={{
+        maxWidth: 'clamp(300px, 50%, 100%)',
+        border: '1px solid hotpink',
+        padding: 8,
+      }}
+    >
+      <ButtonGroup
+        style={{ display: 'flex' }}
```
Comment (bentleyvk) [inline stories/core/Buttons/ButtonGroup.stories.tsx:79]: Not really nice that it doesn't work out of the box and need to change from `inline-flex` to `flex` 😕

File: stories/core/Buttons/ButtonGroup.stories.tsx [hunk 4]
```diff
@@ -37,3 +54,61 @@ export const WithIcons: Story<ButtonGroupProps> = () => {
...
+        overflowButton={(overflowStart) => (
+          <DropdownMenu
+            menuItems={(close) =>
+              Array(buttons.length - overflowStart + 1)
+                .fill(null)
+                .map((_, _index) => {
+                  const index = overflowStart + _index;
+                  const onClick = (close: () => void) => {
```
Comment (bentleyvk) [inline stories/core/Buttons/ButtonGroup.stories.tsx:87]: You don't really need to have `close` as a param.

File: src/core/utils/hooks/useOverflow.test.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import { render } from '@testing-library/react';
+import React from 'react';
+import { useOverflow } from './useOverflow';
+
+const MockComponent = ({ children }: { children: React.ReactNode[] }) => {
+  const [overflowRef, visibleCount] = useOverflow(children);
```
Comment (bentleyvk) [inline src/core/utils/hooks/useOverflow.test.tsx:10]: Test when `disabled` is `true`

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 3 above)
Comment (veekeys) [reply stories/core/Buttons/ButtonGroup.stories.tsx:79]: Agree :O

File: stories/core/Buttons/ButtonGroup.stories.tsx [hunk 6]
```diff
@@ -37,3 +54,61 @@ export const WithIcons: Story<ButtonGroupProps> = () => {
...
+    ));
+
+  return (
+    <div
+      style={{
+        maxWidth: 'clamp(300px, 50%, 100%)',
+        border: '1px solid hotpink',
+        padding: 8,
+      }}
+    >
```
Comment (veekeys) [inline stories/core/Buttons/ButtonGroup.stories.tsx:77]: I would add just a comment above to change the size to see it working or smth..

File: stories/core/Buttons/ButtonGroup.stories.tsx [hunk 7]
```diff
@@ -37,3 +54,61 @@ export const WithIcons: Story<ButtonGroupProps> = () => {
     </ButtonGroup>
   );
 };
+
+export const Overflow: Story<ButtonGroupProps> = (args) => {
+  const buttons = Array(10)
```
Comment (veekeys) [inline stories/core/Buttons/ButtonGroup.stories.tsx:59]: I am thinking.. In real use case there wont be any buttons array. Usually they just drop one by one buttons in there. And if we return index, we kinda force them to change that and put all the buttons in array...

File: stories/core/Buttons/ButtonGroup.stories.tsx [hunk 8]
```diff
@@ -3,25 +3,42 @@
...
   component: ButtonGroup,
+  argTypes: {
+    style: { control: { disable: true } },
+    className: { control: { disable: true } },
+    id: { control: { disable: true } },
+    children: { control: { disable: true } },
+  },
   parameters: {
-    controls: { hideNoControlsWarning: true },
+    creevey: { skip: ['Overflow'] },
```
Comment (veekeys) [inline stories/core/Buttons/ButtonGroup.stories.tsx:35]: Didnt want to bother with a test? Would be nice :D

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 6 above)
Comment (bentleyvk) [reply stories/core/Buttons/ButtonGroup.stories.tsx:77]: I really like Blueprint breadcrumbs demo:
blueprint-breadcrumbs
https://blueprintjs.com/docs/#core/components/breadcrumbs

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 3 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx:79]: Oh yeah, good catch. This needs to be fixed in CSS. I put this here 2 weeks ago and forgot

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 6 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx]: Adding Slider was too much code so I just added text

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 7 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx:57]: If they want to menu items, they need an array somewhere

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 8 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx]: Added

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 4 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx]: ?

File: src/core/ButtonGroup/ButtonGroup.tsx (hunk 2 above)
Comment (mayank99) [reply src/core/ButtonGroup/ButtonGroup.tsx]: Fixed

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 3 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx]: Created https://github.com/iTwin/iTwinUI/pull/247

File: src/core/utils/hooks/useOverflow.test.tsx (hunk 5 above)
Comment (mayank99) [reply src/core/utils/hooks/useOverflow.test.tsx]: Added

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 8 above)
Comment (veekeys) [reply stories/core/Buttons/ButtonGroup.stories.tsx]: I meant with an action to show that overflow.. but ok :D

File: stories/core/Buttons/ButtonGroup.stories.tsx (hunk 3 above)
Comment (mayank99) [reply stories/core/Buttons/ButtonGroup.stories.tsx]: Changed to conditional 100%

**Review by bentleyvk — APPROVED** [review]

**Pull request merged**